source/common/upstream/maglev_lb.cc

```diff
@@ -14,27 +14,48 @@ MaglevTable::MaglevTable(const HostVector& hosts, uint64_t table_size) : table_s
     return;
   }
 
+  // Compute maximum host weight. If this is zero, we are doing unweighted Maglev.
+  uint32_t max_host_weight = 0;
+  for (const auto& host : hosts) {
+    max_host_weight = std::max(host->weight(), max_host_weight);
+  }
+  // Integer range to normalize and scale backend weights to.
+  const uint32_t backend_weight_scale = 65535;
+
   // Implementation of pseudocode listing 1 in the paper (see header file for more info).
   std::vector<TableBuildEntry> table_build_entries;
   table_build_entries.reserve(hosts.size());
   for (const auto& host : hosts) {
     const std::string& address = host->address()->asString();
+    const uint32_t weight =
+        max_host_weight > 0 ? (host->weight() * backend_weight_scale) / max_host_weight : 0;
     table_build_entries.emplace_back(HashUtil::xxHash64(address) % table_size_,
-                                     (HashUtil::xxHash64(address, 1) % (table_size_ - 1)) + 1);
+                                     (HashUtil::xxHash64(address, 1) % (table_size_ - 1)) + 1,
+                                     weight);
   }
 
   table_.resize(table_size_);
   uint64_t table_index = 0;
+  uint32_t iteration = 1;
   while (true) {
     for (uint64_t i = 0; i < hosts.size(); i++) {
-      uint64_t c = permutation(table_build_entries[i]);
+      TableBuildEntry& entry = table_build_entries[i];
+      // Only consider weight if we are doing weighted Maglev.
+      if (max_host_weight) {
+        // Counts are in units of backend_weight_scale;
```

Comment:
Can you add more comments in this section? Since the paper doesn't cover it's not immediately intuitive to me what all of this is doing.

Reply:
Sure. It's actually doing something very similar to what a WRR pick would do, I can comment.